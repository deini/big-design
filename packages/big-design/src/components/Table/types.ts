import { ReactNode } from 'react';

import { MarginProps } from '../../mixins';
import { PaginationProps } from '../Pagination';

export interface TableSelectable<T> {
  itemType?: string;
  selectedItems: T[];
  onSelectionChange(selectedItems: T[]): void;
}

export interface TableItem {
  id?: string | number;
  [key: string]: any;
}

export interface TableColumn<T> {
  render: React.ComponentType<T> | ((props: T & { children?: ReactNode }, context?: any) => string | number);
  align?: 'left' | 'center' | 'right';
  header: string | React.ReactNode;
  verticalAlign?: 'top' | 'center';
  width?: number | string;
  withPadding?: boolean;
}

export type TablePagination = Omit<PaginationProps, keyof MarginProps>;

export interface TableProps<T> extends React.TableHTMLAttributes<HTMLTableElement>, MarginProps {
  columns: Array<TableColumn<T>>;
  items: T[];
  keyField?: string;
  pagination?: TablePagination;
  selectable?: TableSelectable<T>;
  stickyHeader?: boolean;
}
