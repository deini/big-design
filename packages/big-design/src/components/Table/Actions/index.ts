export { Actions, ActionsProps } from './Actions';
