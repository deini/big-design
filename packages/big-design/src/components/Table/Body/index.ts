export { Body, BodyProps } from './Body';
