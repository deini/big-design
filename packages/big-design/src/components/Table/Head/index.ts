export { Head, HeadProps } from './Head';
