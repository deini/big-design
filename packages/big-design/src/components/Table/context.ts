import React from 'react';

type SectionContext = 'thead' | 'tbody' | 'tfoot';

export const TableSectionContext = React.createContext<SectionContext>('tbody');

interface Context {
  stickyHeader?: boolean;
  tableId: string;
}

export const TableContext = React.createContext<Context>({
  tableId: '',
});
