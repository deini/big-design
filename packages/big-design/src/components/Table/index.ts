export { Table, TableFigure } from './Table';
export { TableProps } from './types';
