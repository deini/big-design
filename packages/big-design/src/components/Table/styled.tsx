import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { TableProps } from './types';

export const StyledTableFigure = styled.figure`
  margin: 0;
  margin-bottom: ${({ theme }) => `${theme.spacing.xLarge}`};
  max-width: 100%;
  overflow-x: auto;
  position: relative;
  white-space: nowrap;

  ${({ theme }) => theme.breakpoints.tablet} {
    white-space: normal;
  }
`;

export const StyledTable = styled.table<Partial<TableProps>>`
  border-collapse: collapse;
  border-color: transparent;
  color: ${({ theme }) => theme.colors.secondary70};
  text-align: left;
  width: 100%;
`;

StyledTableFigure.defaultProps = { theme: defaultTheme };
StyledTable.defaultProps = { theme: defaultTheme };
