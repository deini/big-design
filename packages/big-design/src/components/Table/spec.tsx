import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import React, { CSSProperties } from 'react';

import { Table } from './Table';

interface SimpleTableOptions {
  className?: string;
  dataTestId?: string;
  id?: string;
  style?: CSSProperties;
  columns?: any[];
}

const getSimpleTable = ({ className, columns, dataTestId, id, style }: SimpleTableOptions = {}) => (
  <Table
    id={id}
    data-testid={dataTestId}
    className={className}
    style={style}
    columns={
      columns || [
        { header: 'Sku', render: ({ sku }) => sku },
        { header: 'Name', render: ({ name }) => name },
        { header: 'Stock', render: ({ stock }) => stock },
      ]
    }
    items={[
      { sku: 'SM13', name: '[Sample] Smith Journal 13', stock: 25 },
      { sku: 'DPB', name: '[Sample] Dustpan & Brush', stock: 34 },
      { sku: 'OFSUC', name: '[Sample] Utility Caddy', stock: 45 },
      { sku: 'CLC', name: '[Sample] Canvas Laundry Cart', stock: 2 },
      { sku: 'CGLD', name: '[Sample] Laundry Detergent', stock: 29 },
    ]}
  />
);

test('renders a simple table', () => {
  const { container } = render(getSimpleTable());

  expect(container.firstChild).toMatchSnapshot();
});

test('generates a table id automatically', () => {
  const { getByRole } = render(getSimpleTable());

  const table = getByRole('table');

  expect(table.id).toBeTruthy();
});

test('forwards id and testid when provided', () => {
  const id = 'testId';
  const dataTestId = 'dataTestId';

  const { getByRole } = render(getSimpleTable({ id, dataTestId }));

  const table = getByRole('table');

  expect(table.id).toBe(id);
  expect(table.dataset.testid).toBe(dataTestId);
});

test('does not forward styles', () => {
  const { container } = render(getSimpleTable({ className: 'test', style: { background: 'red' } }));

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('renders column with custom component', () => {
  const { getAllByTestId } = render(
    getSimpleTable({
      columns: [
        { header: 'Sku', render: ({ sku }: any) => sku },
        { header: 'Name', render: ({ name }: any) => <h3 data-testid="name">{name}</h3> },
      ],
    }),
  );

  expect(getAllByTestId('name').length).toBe(5);
});

test('tweaks column styles with props', () => {
  const { container } = render(
    getSimpleTable({
      columns: [
        {
          header: 'Sku',
          render: ({ sku }: any) => sku,
          align: 'right',
          verticalAlign: 'center',
        },
        {
          header: 'Name',
          render: ({ name }: any) => name,
          width: 100,
          withPadding: false,
        },
      ],
    }),
  );

  const headers = container.querySelectorAll('th');
  const skuHeader = headers[0];
  const nameHeader = headers[1];

  expect(skuHeader).toHaveStyle('text-align: right');
  expect(skuHeader).not.toHaveStyle('vertical-align: center');

  expect(nameHeader).toHaveStyle('width: 100px');
  expect(skuHeader).not.toHaveStyle('padding: 0');

  const columns = container.querySelectorAll('tbody td');
  const skuTd = columns[0];
  const nameTd = columns[1];

  expect(skuTd).toHaveStyle(`
    text-align: right;
    vertical-align: center;
  `);

  expect(nameTd).toHaveStyle(`
    width: 100px;
    padding: 0;
  `);
});

test('renders a pagination component', () => {
  const onItemsPerPageChange = jest.fn();
  const onPageChange = jest.fn();

  const { container, getByTitle } = render(
    <Table
      columns={[
        { header: 'Sku', render: ({ sku }) => sku },
        { header: 'Name', render: ({ name }) => name },
        { header: 'Stock', render: ({ stock }) => stock },
      ]}
      items={[
        { sku: 'SM13', name: '[Sample] Smith Journal 13', stock: 25 },
        { sku: 'DPB', name: '[Sample] Dustpan & Brush', stock: 34 },
        { sku: 'OFSUC', name: '[Sample] Utility Caddy', stock: 45 },
        { sku: 'CLC', name: '[Sample] Canvas Laundry Cart', stock: 2 },
        { sku: 'CGLD', name: '[Sample] Laundry Detergent', stock: 29 },
      ]}
      pagination={{
        currentPage: 1,
        itemsPerPage: 3,
        totalItems: 5,
        itemsPerPageOptions: [3, 5, 10],
        onItemsPerPageChange,
        onPageChange,
      }}
    />,
  );

  fireEvent.click(getByTitle('Next page'));

  expect(onPageChange).toHaveBeenCalledWith(2);
  expect(container.firstChild).toMatchSnapshot();
});

describe('selectable', () => {
  let columns: any;
  let items: any;
  let onSelectionChange: jest.Mock;
  const itemType = 'Product';

  beforeEach(() => {
    onSelectionChange = jest.fn();
    items = [
      { sku: 'SM13', name: '[Sample] Smith Journal 13', stock: 25 },
      { sku: 'DPB', name: '[Sample] Dustpan & Brush', stock: 34 },
      { sku: 'OFSUC', name: '[Sample] Utility Caddy', stock: 45 },
      { sku: 'CLC', name: '[Sample] Canvas Laundry Cart', stock: 2 },
      { sku: 'CGLD', name: '[Sample] Laundry Detergent', stock: 29 },
    ];
    columns = [
      { header: 'Sku', render: ({ sku }: any) => sku },
      { header: 'Name', render: ({ name }: any) => name },
      { header: 'Stock', render: ({ stock }: any) => stock },
    ];
  });

  test('renders selectable actions and checkboxes', () => {
    const { container, getAllByRole } = render(
      <Table
        columns={columns}
        items={items}
        selectable={{
          itemType,
          onSelectionChange,
          selectedItems: [],
        }}
      />,
    );

    // One per item + Actions (select all) checkbox
    expect(getAllByRole('checkbox').length).toBe(items.length + 1);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('click on select all should call selectedItems with all items', () => {
    const { getAllByRole } = render(
      <Table
        columns={columns}
        items={items}
        selectable={{
          itemType,
          onSelectionChange,
          selectedItems: [],
        }}
      />,
    );

    const [selectAllCheckbox] = getAllByRole('checkbox') as HTMLInputElement[];

    // Select All
    expect(selectAllCheckbox.checked).toBe(false);
    fireEvent.click(selectAllCheckbox);
    expect(onSelectionChange).toHaveBeenCalledWith(items);
  });

  test('select all when already all selected should deselect all items', () => {
    const { getAllByRole } = render(
      <Table
        columns={columns}
        items={items}
        selectable={{
          itemType,
          onSelectionChange,
          selectedItems: items,
        }}
      />,
    );
    const [selectAllCheckbox] = getAllByRole('checkbox') as HTMLInputElement[];

    // Deselect all
    expect(selectAllCheckbox.checked).toBe(true);
    fireEvent.click(selectAllCheckbox);
    expect(onSelectionChange).toHaveBeenCalledWith([]);
  });
});
