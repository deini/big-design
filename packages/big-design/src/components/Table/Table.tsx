import React, { memo, useRef } from 'react';

import { uniqueId } from '../../utils';

import { TableContext } from './context';
import { StyledTable, StyledTableFigure } from './styled';
import { TableItem, TableProps } from './types';
import { Actions } from './Actions';
import { Body } from './Body';
import { Cell } from './Cell';
import { Head } from './Head';
import { Item } from './Item';

export const Table = <T extends TableItem>(props: TableProps<T>): React.ReactElement<TableProps<T>> => {
  const {
    className,
    stickyHeader,
    style,
    items,
    columns,
    pagination,
    selectable,
    id,
    keyField = 'id',
    ...rest
  } = props;
  const tableIdRef = useRef(id || uniqueId('table_'));
  const isSelectable = Boolean(selectable);

  const getItemKey = (item: T, index: number): string | number => {
    if (item[keyField] !== undefined) {
      return item[keyField];
    }

    return index;
  };

  const isItemSelected = (item: T) => {
    return selectable && selectable.selectedItems.includes(item);
  };

  const onItemSelect = (item: T, isSelected: boolean) => {
    if (!selectable) {
      return;
    }

    const { selectedItems, onSelectionChange } = selectable;

    if (isSelected) {
      onSelectionChange([...selectedItems, item]);
    } else {
      onSelectionChange(selectedItems.filter(selectedItem => selectedItem !== item));
    }
  };

  const shouldRenderActions = () => {
    return Boolean(pagination) || Boolean(selectable);
  };

  const renderHeaders = () => (
    <Head>
      <Item isSelectable={isSelectable}>
        {columns.map(({ header, align, width }, index) => (
          <Cell key={index} align={align} width={width}>
            {header}
          </Cell>
        ))}
      </Item>
    </Head>
  );

  const renderItems = () => (
    <Body>
      {items.map((item: T, index) => (
        <Item
          isSelectable={isSelectable}
          key={getItemKey(item, index)}
          onItemSelect={nextValue => onItemSelect(item, nextValue)}
          selected={isItemSelected(item)}
        >
          {props.columns.map(
            ({ render: CellContent, align, verticalAlign, width, withPadding = true }, columnIndex) => (
              <Cell
                key={columnIndex}
                align={align}
                verticalAlign={verticalAlign}
                width={width}
                withPadding={withPadding}
              >
                {/* https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20544 */}
                {/* 
                // @ts-ignore */}
                <CellContent {...item} />
              </Cell>
            ),
          )}
        </Item>
      ))}
    </Body>
  );

  return (
    <TableContext.Provider value={{ stickyHeader }}>
      {shouldRenderActions() && (
        <Actions pagination={pagination} selectable={selectable} items={items} tableId={tableIdRef.current} />
      )}
      <StyledTable {...rest} id={tableIdRef.current}>
        {renderHeaders()}
        {renderItems()}
      </StyledTable>
    </TableContext.Provider>
  );
};

export const TableFigure: React.FC = memo(props => <StyledTableFigure {...props} />);
