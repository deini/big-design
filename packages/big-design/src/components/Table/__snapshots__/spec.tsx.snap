// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a pagination component 1`] = `
.c8 {
  vertical-align: middle;
  height: 2rem;
  width: 2rem;
}

.c12 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  padding: 0.75rem;
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
}

.c0 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

.c9 {
  border-radius: 0.25rem;
  box-shadow: 0px 1px 6px rgba(49,52,64,0.2);
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  background-color: #FFFFFF;
  color: #313440;
  display: inline-block;
  margin: 0;
  max-height: 15.625rem;
  outline: none;
  overflow-y: scroll;
  padding: 0.5rem 0;
  z-index: 1070;
}

.c10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  cursor: default;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.25rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-width: 16rem;
  outline: none;
  padding: 0 1rem;
}

.c10[aria-selected='true'] {
  font-weight: 600;
}

.c10[data-highlighted='true'] {
  background-color: #F6F7FC;
}

.c10 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #313440;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  margin: 0 -1rem;
  padding: 0 1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  pointer-events: auto;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  padding-right: 0.5rem;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c6:focus {
  outline: none;
}

.c6[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c6 + .bd-button {
  margin-top: 0.5rem;
}

.c6:active {
  background-color: #DBE3FE;
}

.c6:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c6:hover:not(:active) {
  background-color: #F0F3FF;
}

.c6[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c11 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  pointer-events: auto;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c11:focus {
  outline: none;
}

.c11[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c11 + .bd-button {
  margin-top: 0.5rem;
}

.c11:active {
  background-color: #DBE3FE;
}

.c11:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c11:hover:not(:active) {
  background-color: #F0F3FF;
}

.c11[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c7 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
  visibility: visible;
}

.c5 {
  color: #313440;
  width: auto;
}

@media (min-width:0px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:0px) {
  .c4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c4 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:720px) {
  .c6 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    width: auto;
  }
}

@media (min-width:720px) {
  .c11 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c11 {
    width: auto;
  }
}

<div
  aria-controls="table_0"
  class="c0 c1"
>
  <div
    class="c2 c3"
    style="margin-left: auto;"
  >
    <div
      aria-label="pagination"
      class="c4 c3"
      role="navigation"
    >
      <div
        class="c2 c3"
      >
        <button
          aria-haspopup="true"
          class="c5 c6"
          id="trigger_2"
          role="button"
          tabindex="0"
        >
          <span
            class="c7"
          >
            1 - 3 of 5
            <svg
              class="c8"
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24"
            >
              <title />
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M8.71 11.71l2.59 2.59c.39.39 1.02.39 1.41 0l2.59-2.59c.63-.63.18-1.71-.71-1.71H9.41c-.89 0-1.33 1.08-.7 1.71z"
              />
            </svg>
          </span>
        </button>
        <ul
          aria-labelledby="trigger_2"
          class="c9"
          id="dropdown_1"
          role="listbox"
          style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
          tabindex="-1"
        >
          <li
            class="c10"
            data-value="3"
            id="dropdown_1-item-0"
            role="option"
            tabindex="-1"
          >
            3
             per page
          </li>
          <li
            class="c10"
            data-value="5"
            id="dropdown_1-item-1"
            role="option"
            tabindex="-1"
          >
            5
             per page
          </li>
          <li
            class="c10"
            data-value="10"
            id="dropdown_1-item-2"
            role="option"
            tabindex="-1"
          >
            10
             per page
          </li>
        </ul>
      </div>
      <div
        class="c2 c3"
      >
        <button
          class="c5 c11"
          disabled=""
          role="button"
          tabindex="0"
        >
          <span
            class="c7"
          >
            <svg
              class="c12"
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24"
            >
              <title>
                Previous page
              </title>
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M14.71 6.71a.996.996 0 00-1.41 0L8.71 11.3a.996.996 0 000 1.41l4.59 4.59a.996.996 0 101.41-1.41L10.83 12l3.88-3.88c.39-.39.38-1.03 0-1.41z"
              />
            </svg>
          </span>
        </button>
        <button
          class="c5 c11"
          role="button"
          tabindex="0"
        >
          <span
            class="c7"
          >
            <svg
              class="c12"
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24"
            >
              <title>
                Next page
              </title>
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M9.29 6.71a.996.996 0 000 1.41L13.17 12l-3.88 3.88a.996.996 0 101.41 1.41l4.59-4.59a.996.996 0 000-1.41L10.7 6.7c-.38-.38-1.02-.38-1.41.01z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a simple table 1`] = `
.c0 {
  border-collapse: collapse;
  border-color: transparent;
  color: #313440;
  text-align: left;
  width: 100%;
}

.c3 > tr {
  border-bottom: 1px solid #D9DCE9;
}

.c2 {
  box-sizing: border-box;
  font-size: 1rem;
  padding: 0.75rem;
  background-color: #F6F7FC;
  box-shadow: inset 0px -1px 0px #D9DCE9,inset 0px 1px 0px #D9DCE9;
  color: #5E637A;
  white-space: nowrap;
}

.c4 {
  box-sizing: border-box;
  font-size: 1rem;
  padding: 0.75rem;
  color: #313440;
  padding: 0.75rem;
}

.c1 {
  background-color: transparent;
}

.c1:hover {
  background-color: #F6F7FC;
}

<table
  class="c0"
  id="table_0"
>
  <thead
    class=""
  >
    <tr
      class="c1"
    >
      <th
        class="c2"
      >
        Sku
      </th>
      <th
        class="c2"
      >
        Name
      </th>
      <th
        class="c2"
      >
        Stock
      </th>
    </tr>
  </thead>
  <tbody
    class="c3"
  >
    <tr
      class="c1"
    >
      <td
        class="c4"
      >
        SM13
      </td>
      <td
        class="c4"
      >
        [Sample] Smith Journal 13
      </td>
      <td
        class="c4"
      >
        25
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c4"
      >
        DPB
      </td>
      <td
        class="c4"
      >
        [Sample] Dustpan & Brush
      </td>
      <td
        class="c4"
      >
        34
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c4"
      >
        OFSUC
      </td>
      <td
        class="c4"
      >
        [Sample] Utility Caddy
      </td>
      <td
        class="c4"
      >
        45
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c4"
      >
        CLC
      </td>
      <td
        class="c4"
      >
        [Sample] Canvas Laundry Cart
      </td>
      <td
        class="c4"
      >
        2
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c4"
      >
        CGLD
      </td>
      <td
        class="c4"
      >
        [Sample] Laundry Detergent
      </td>
      <td
        class="c4"
      >
        29
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`selectable renders selectable actions and checkboxes 1`] = `
.c9 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c10 {
  color: #313440;
  margin: 0 0 1rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5rem;
  margin-left: 0.75rem;
}

.c10:last-child {
  margin-bottom: 0;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF;
  border: 1px solid #D9DCE9;
  border-color: #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 1.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1.25rem;
}

.c6:focus + .c8 {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c8 svg {
  visibility: hidden;
}

.c1 {
  padding: 0.75rem;
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
}

.c0 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

@media (min-width:0px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:0px) {
  .c4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c4 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div
  aria-controls="table_0"
  class="c0 c1"
>
  <div
    class="c2 c3"
  >
    <div
      class="c4 c3"
    >
      <div
        class="c5"
      >
        <input
          aria-labelledby="checkBox_label_2"
          class="c6 c7"
          id="checkBox_1"
          type="checkbox"
        />
        <label
          aria-hidden="true"
          class="c8"
          for="checkBox_1"
        >
          <svg
            class="c9"
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
          >
            <title />
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M9 16.17L5.53 12.7a.996.996 0 10-1.41 1.41l4.18 4.18c.39.39 1.02.39 1.41 0L20.29 7.71a.996.996 0 10-1.41-1.41L9 16.17z"
            />
          </svg>
        </label>
      </div>
      <p
        class="c10"
      >
        0
        /
        5
         
        Product
      </p>
    </div>
  </div>
</div>
`;
